src/OrleansRuntime/Storage/MemoryStorageGrain.cs

```diff
@@ -66,57 +63,67 @@ private GrainStateStore GetStoreForGrain(string grainType)
         {
             private Logger logger;
-            private readonly IDictionary<string, Tuple<IDictionary<string, object>, long>> grainStateStorage = new Dictionary<string, Tuple<IDictionary<string, object>, long>>();
-            
             public GrainStateStore(Logger logger)
             {
                 this.logger = logger;
             }
+            private readonly IDictionary<string, IGrainState> grainStateStorage = new Dictionary<string, IGrainState>();
 
-            public Tuple<IDictionary<string, object>, string> GetGrainState(string grainId)
+            public IGrainState GetGrainState(string grainId)
             {
-                Tuple<IDictionary<string, object>, long> state;
-                if(grainStateStorage.TryGetValue(grainId, out state))
-                    return Tuple.Create(state.Item1, state.Item2.ToString());
-                else
-                    return Tuple.Create<IDictionary<string, object>, string>(null, null); // upon first read, return null/invalid etag, to mimic Azure Storage.
+                IGrainState entry;
+                grainStateStorage.TryGetValue(grainId, out entry);
+                return entry;
             }
 
-            public string UpdateGrainState(string grainStoreKey, IDictionary<string, object> state, string receivedEtag)
+            public string UpdateGrainState(string grainId, IGrainState grainState)
             {
-                long currentETag = 0;
-                Tuple<IDictionary<string, object>, long> oldState;
-                if (grainStateStorage.TryGetValue(grainStoreKey, out oldState)) {
-                    currentETag = oldState.Item2;
+                IGrainState entry;
+                grainStateStorage.TryGetValue(grainId, out entry);
+                if (entry == null)
+                {
+                    grainStateStorage[grainId] = grainState;
+                    return grainState.ETag;
                 }
-                ValidateEtag(currentETag, receivedEtag, grainStoreKey, "Update");
-                currentETag++;
-                grainStateStorage[grainStoreKey] = Tuple.Create(state, currentETag);
-                return currentETag.ToString();
+
+                ValidateEtag(grainState.ETag, entry.ETag, grainId, "Update");
+
+                grainState.ETag = NewEtag();
+                grainStateStorage[grainId] = grainState;
+                return grainState.ETag;
             }
 
-            public void DeleteGrainState(string grainStoreKey, string receivedEtag)
+            public string DeleteGrainState(string grainId, string eTag)
             {
-                long currentETag = 0;
-                Tuple<IDictionary<string, object>, long> oldState;
-                if (grainStateStorage.TryGetValue(grainStoreKey, out oldState)){
-                    currentETag = oldState.Item2;
+                IGrainState entry;
+                grainStateStorage.TryGetValue(grainId, out entry);
+                if (entry == null)
```

Comment:
same big here as in update: `if (entry == null)` is NOT a valid case here, should throw etag violation.